src/daggy.js

```diff
@@ -1,127 +1,136 @@
-const getInstance = require('fantasy-helpers/src/get-instance');
-const {constant} = require('fantasy-combinators');
-
-/**
-  ## `daggy.tagged(arguments)`
-
-  Creates a new constructor with the given field names as
-  arguments and properties. Allows `instanceof` checks with
-  returned constructor.
-
-  ```javascript
-  const Tuple3 = daggy.tagged('x', 'y', 'z');
-
-  const _123 = Tuple3(1, 2, 3); // optional new keyword
-  _123.x == 1 && _123.y == 2 && _123.z == 3; // true
-  _123 instanceof Tuple3; // true
-  ```
-**/
-function tagged() {
-    const fields = [].slice.apply(arguments);
-
-    function toString(args) {
-      const x = [].slice.apply(args);
-      return () => {
-        const values = x.map((y) => y.toString());
-        return '(' + values.join(', ') + ')';
-      };
+const { toString } = require('sanctuary-type-classes')
+const type = require('sanctuary-type-identifiers')
+
+// Name of prop used to store:
+// * name of variant of a sum type
+const TAG = '@@tag'
+// * array of arguments used to create a value (to speed up `cata`)
+const VALUES = '@@values'
+
+// adopted version of withValue from  https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty
+const defProp = (obj, prop, val) => {
+  var desc = defProp.desc || (
+    defProp.desc = {
+      enumerable: false,
+      writable: false,
+      configurable: false,
+      value: null
     }
+  )
+  desc.value = val
+  Object.defineProperty(obj, prop, desc)
+}
 
-    function wrapped() {
-        const self = getInstance(this, wrapped);
-        var i;
-
-        if(arguments.length != fields.length)
-            throw new TypeError('Expected ' + fields.length + ' arguments, got ' + arguments.length);
-
-        for(i = 0; i < fields.length; i++)
-            self[fields[i]] = arguments[i];
+// optimised version of `arr.map(toString).join(', ')`
+const arrToString = (arr) => {
+  if (arr.length === 0) {
+    return ''
+  }
+  let str = '(' + toString(arr[0])
+  for (var i = 1; i < arr.length; i++) {
+    str += ', ' + toString(arr[i])
```

Comment:
The `+=` operator may cost a lot of performance here compared to `str = str + ...`. This has been optimized in V8 v5.6, but Node 7 still uses v5.4, and the issue will remain in all older version of Node.